rfd/0053-passwordless-fido2.md

```diff
@@ -0,0 +1,494 @@
+All user<->browser interaction steps depend on the browser implementation. The
+observations below are meant to inform RFD readers on the status quo at the time
+of writing.
+
+(1) may be skipped if there is only a single authenticator present in the
+machine. In case of ambiguity (more than one authenticator), then the user is
+prompted to pick their authenticator of choice by tapping it.
+
+(2) and (3) may be skipped for biometric authenticators - the initial tap is
+sufficient for both verification and presence checks.
+
+(4) may also be skipped if there is a single resident key for the RPID (Relying
+Party ID) in the authenticator, otherwise a menu with options shows up.
+
+### CLI clients (aka `tsh`)
+
+There are essentially two options for CLI clients / `tsh`:
+
+1. Leverage browsers and delegate FIDO2 authentication to them; or
+2. Write CLI-native FIDO authentication.
+
+(1) is attractive enough to merit discussion. Its main downside is that we can't
+get a purely CLI-based flow, a browser needs to pop up. The upside is that we
+get to benefit from the work browsers do for authentication and security. This
+option means we get support for FIDO2 and native biometrics in a single stroke,
+without having to deal with the delicate issues they pose ourselves.
+
+(2) is the current approach used by Teleport, as it brings characteristics that
+are important to our users. Unfortunately, it forces us to tackle some of the
+difficult problems already solved by browsers, which often require
+platform-specific work as well.
+
+A compromise between both approaches is possible. For example, `tsh` could rely
+on browsers in most cases (1), but implement a best-effort CLI-native solution
+(2). This takes pressure out of the CLI implementation and allows it to focus in
+the more impactful use-cases, instead of having to provide full functionality.
+
+For the moment, the RFD chooses the CLI-native approach (2). A possible design
+for browser-based CLI authentication was briefly considered, but discarded due
+to security concerns.
+
+#### CLI-native authentication
+
+CLI-native passwordless relies on
+[libfido2](https://github.com/Yubico/libfido2), a C library that allows
+interaction with FIDO2 resident keys. (The official Go wrapper is
+[github.com/keys-pub/go-libfido2](https://github.com/keys-pub/go-libfido2/).)
+
+One of the issues for passwordless interaction is that the client (`tsh`) gets
+no information about registered credentials. The usual tactic of attempting an
+assertion on all hardware keys falls short, as an authenticator may hold
+multiple keys for the same Relying Party - in this case, the authenticator picks
+the account, not the user!
+
+The solution is similar to the one used by browsers - we make all hardware keys
+blink and ask the user to choose the desired key by touching it. Authentication
+proceeds by enumerating the credentials from that key and letting the user,
+again, pick the desired one. (See libfido2's [examples/select.c](
```

Comment:
what do these credentials look like, are they readable enough that it's not completely unreasonable to make the user pick?

Reply:
They have human-readable field specifically for this purpose (ie the Teleport username), so should be fine.

https://www.w3.org/TR/webauthn-2/#dictionary-user-credential-params